### PR TITLE
modularize inferencing during and after training, Stable Diffusion(SD)

### PR DESCRIPTION
**Pull request by charchit7**
# What does this PR do?

Part of #6545 for SD

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sayakpaul  

I think I should have kept the `args.validation_prompts` while modifying the code. But forgot with the flow. Please let me know if it's required.

### CODE REVIEW

File: examples/text_to_image/train_text_to_image.py [hunk 1]
```diff
@@ -71,7 +72,7 @@ def save_model_card(
 ):
     img_str = ""
     if len(images) > 0:
-        image_grid = make_image_grid(images, 1, len(args.validation_prompts))
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:74]: It cannot be like that.

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -92,8 +93,8 @@ def save_model_card(
     model_card = f"""
 # Text-to-image finetuning - {repo_id}
 
-This pipeline was finetuned from **{args.pretrained_model_name_or_path}** on the **{args.dataset_name}** dataset. Below are some example images generated with the finetuned pipeline using the following prompts: {args.validation_prompts}: \n
-{img_str}
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:96]: Let's not touch these things please.

File: examples/text_to_image/train_text_to_image.py [hunk 3]
```diff
@@ -104,7 +105,7 @@ def save_model_card(
 import torch
 
 pipeline = DiffusionPipeline.from_pretrained("{repo_id}", torch_dtype=torch.float16)
-prompt = "{args.validation_prompts[0]}"
+prompt = "Baby Yoda in the wild"
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:108]: Same.

File: examples/text_to_image/train_text_to_image.py [hunk 4]
```diff
@@ -138,7 +139,9 @@ def save_model_card(
         f.write(yaml + model_card)
 
 
-def log_validation(vae, text_encoder, tokenizer, unet, args, accelerator, weight_dtype, epoch):
+def log_validation(
+    vae, args, text_encoder, tokenizer, accelerator, weight_dtype, epoch, unet=None, is_final_validation=False
+):
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:144]:
```suggestion
def log_validation(
    unet, vae, text_encoder, tokenizer, args, accelerator, weight_dtype, epoch, is_final_validation=False
):
```

Let's keep the signature mentally more readable.

File: examples/text_to_image/train_text_to_image.py [hunk 5]
```diff
@@ -163,33 +166,57 @@ def log_validation(vae, text_encoder, tokenizer, unet, args, accelerator, weight
...
-    images = []
-    for i in range(len(args.validation_prompts)):
-        with torch.autocast("cuda"):
-            image = pipeline(args.validation_prompts[i], num_inference_steps=20, generator=generator).images[0]
+    validation_prompts = [
+        "cute Elon Musk character",
+        "A pokemon with wings",
+        "a photo of yoda",
+        "a cute creature with blue wings",
+    ]
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:174]: Shouldn't be like this.

File: examples/text_to_image/train_text_to_image.py [hunk 6]
```diff
@@ -163,33 +166,57 @@ def log_validation(vae, text_encoder, tokenizer, unet, args, accelerator, weight
...
             logger.warn(f"image logging not implemented for {tracker.name}")
 
-    del pipeline
-    torch.cuda.empty_cache()
+        del pipeline
+        gc.collect()
+        torch.cuda.empty_cache()
 
-    return images
+        return image_logs
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:219]: Why are these inside one more indentation?

**Review by sayakpaul — COMMENTED** [review]
Thanks for initiating this. 

However, there some critical things we need to address first. 

There are two instances of intermediate validation which must not be the case:

* https://github.com/charchit7/diffusers/blob/dcd201015a868ad94b27002a6bb768cbe812c0e0/examples/text_to_image/train_text_to_image.py#L1038
* https://github.com/charchit7/diffusers/blob/dcd201015a868ad94b27002a6bb768cbe812c0e0/examples/text_to_image/train_text_to_image.py#L1057
* https://github.com/charchit7/diffusers/blob/dcd201015a868ad94b27002a6bb768cbe812c0e0/examples/text_to_image/train_text_to_image.py#L1038 -- this one doesn't consider the proper EMA'd UNet

Then while doing the final inference [here](https://github.com/charchit7/diffusers/blob/dcd201015a868ad94b27002a6bb768cbe812c0e0/examples/text_to_image/train_text_to_image.py#L1106), we shouldn't be passing the UNet, text encoder, etc. We should just set them to `None` as we're serializing the pipeline already with `.save_pretrained()`. 

This is where we leverage `is_final_validation` and upon detecting that to be True, we initialize the pipeline within `log_validation()` like so:

```python
pipeline = StableDiffusionPipeline.from_pretrained(args.output_dir, ...)
```

File: examples/text_to_image/train_text_to_image.py (hunk 5 above)
Comment (charchit7) [reply examples/text_to_image/train_text_to_image.py:174]: I should take from arguments, right? `args.validation_prompts`

**Comment by charchit7** [comment]
@sayakpaul thank you so much for detailed review. I'll address them as soon as I can.

File: examples/text_to_image/train_text_to_image.py (hunk 6 above)
Comment (charchit7) [reply examples/text_to_image/train_text_to_image.py:219]: sleep deprived coding. 😅

**Comment by charchit7** [comment]
Hi @sayakpaul sorry for the delay. Please check the PR if I am missing something when you are free.

**Comment by sayakpaul** [comment]
Can you provide an example command with which I can test it?

**Comment by charchit7** [comment]
Hi @sayakpaul `Pytorch Example CPU tests` are failing due passing unet also in the `log_validation` later
pytest results: 
image

**Comment by sayakpaul** [comment]
You need to take care of that before we can merge it, though. And also a working training command so that I can test it.

**Pull request closed without merging**